File: TimeCalculator/TimeLog.cs
namespace TimeCalculator;

public class TimeLog
{
    private readonly IList<TimeInterval> _timeIntervals;

    /// <summary>
    /// Initialize a TimeLog object with a list of TimeIntervals.
    /// </summary>
    /// <param name="timeIntervals"></param>
    public TimeLog(IList<TimeInterval> timeIntervals)
    {
        _timeIntervals = timeIntervals;
    }

    /// <summary>
    /// Calculate the total time in minutes.
    /// </summary>
    private int GetTotalTimeInMinutes()
    {
        var totalTimeInMinutes = 0;
        foreach (var timeInterval in _timeIntervals)
            totalTimeInMinutes += timeInterval.GetTimeIntervalInMinutes();
        return totalTimeInMinutes;
    }

    /// <summary>
    /// Get the total time in the format "hh:mm".
    /// </summary>
    /// <returns></returns>
    public string GetTotalTimeString()
    {
        var totalTimeInMinutes = GetTotalTimeInMinutes();
        var hours = totalTimeInMinutes / 60;
        var minutes = totalTimeInMinutes % 60;
        return $"{hours, 2}:{minutes, 2}";
    }

    /// <summary>
    /// format with 1. From 12:10 to 13:10
    /// </summary>
    /// <returns></returns>
    public string GetTimeLogStrings()
    {
        var timeLogStrings = new List<string>();
        foreach (var timeInterval in _timeIntervals)
            timeLogStrings.Add(timeInterval.GetTimeIntervalString());
        return string.Join("\n", timeLogStrings);
    }
}

File: src/time_log.rs
use super::time_interval::TimeInterval;

pub struct TimeLog {
    time_intervals: Vec<TimeInterval>,
}

impl TimeLog {
    pub fn new(time_intervals: Vec<TimeInterval>) -> Self {
        Self { time_intervals }
    }

    pub fn get_total_time_in_minutes(&self) -> u16 {
        self.time_intervals
            .iter()
            .fold(0, |acc, x| acc + x.get_time_interval_in_minutes())
    }

    pub fn get_total_time_string(&self) -> String {
        let total = self.get_total_time_in_minutes();
        let hours = total / 60;
        let minutes = total % 60;
        format!("{:02}:{:02}", hours, minutes)
    }

    pub fn get_time_log_strings(&self) -> String {
        self.time_intervals
            .iter()
            .map(|x| x.get_time_interval_string())
            .reduce(|mut acc, x| {
                acc.push_str(&x);
                acc
            })
            .unwrap()
    }
}
